### ISSUE TITLE
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.IllegalArgumentException: Invalid factoryId!

### ISSUE THREAD

**Issue by dsukhoroslov** [hazelcast/hazelcast#6276]
Under some circumstancies I see the following exceptions in the server logs:

```
2015-09-21 00:25:32.935 [hz.default.client.thread-4] ERROR com.hazelcast.client.ClientEngine - [192.168.1.100]:10501 [default] [3.5.2] java.lang.IllegalArgumentException: Invalid factoryId! Expected: 2, Current: -1
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.IllegalArgumentException: Invalid factoryId! Expected: 2, Current: -1
	at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:380) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:282) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.loadRequest(ClientEngineImpl.java:396) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.run(ClientEngineImpl.java:371) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_21]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_21]
	at java.lang.Thread.run(Thread.java:722) [na:1.7.0_21]
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) [hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92) [hazelcast-all-3.5.2.jar:3.5.2]
Caused by: java.lang.IllegalArgumentException: Invalid factoryId! Expected: 2, Current: -1
	at com.hazelcast.nio.serialization.DefaultPortableReader.checkFactoryAndClass(DefaultPortableReader.java:271) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.DefaultPortableReader.readPortable(DefaultPortableReader.java:223) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.client.AuthenticationRequest.read(AuthenticationRequest.java:213) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.client.ClientRequest.readPortable(ClientRequest.java:95) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:84) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:73) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:29) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:276) ~[hazelcast-all-3.5.2.jar:3.5.2]
	... 7 common frames omitted
```

or the opposite one:

```
2015-09-22 08:49:14.810 [hz.default.client.thread-26] ERROR com.hazelcast.client.ClientEngine - [192.168.1.100]:10501 [default] [3.5.2] java.lang.IllegalArgumentException: Invalid factoryId! Expected: -1, Current: 2
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.IllegalArgumentException: Invalid factoryId! Expected: -1, Current: 2
	at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:380) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:282) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.loadRequest(ClientEngineImpl.java:396) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.run(ClientEngineImpl.java:371) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_21]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_21]
	at java.lang.Thread.run(Thread.java:722) [na:1.7.0_21]
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) [hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92) [hazelcast-all-3.5.2.jar:3.5.2]
Caused by: java.lang.IllegalArgumentException: Invalid factoryId! Expected: -1, Current: 2
	at com.hazelcast.nio.serialization.DefaultPortableReader.checkFactoryAndClass(DefaultPortableReader.java:271) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.DefaultPortableReader.readPortable(DefaultPortableReader.java:223) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.client.AuthenticationRequest.read(AuthenticationRequest.java:213) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.client.impl.client.ClientRequest.readPortable(ClientRequest.java:95) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:84) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:73) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:29) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41) ~[hazelcast-all-3.5.2.jar:3.5.2]
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:276) ~[hazelcast-all-3.5.2.jar:3.5.2]
	... 7 common frames omitted
```

I do have my own PortableSerializationFactory with id 2, so it is not clear why it causes any issues? The exception thown quite rare, but when it happens the system just hangs.

**Comment by asimarslan** [comment]
Hi @dsukhoroslov, 
Can you provide more detail on the issue to reproduce the case? If possible you can try the 3.6-EA
Thanks.

**Comment by asimarslan** [comment]
@dsukhoroslov, I'm closing the issue as we couldn't reproduce it. Please reopen if you can provide more info to reproduce it. Thanks

**Comment by dsukhoroslov** [comment]
Hi Asim,

Ok, let's close it. I can not reproduce it too, may be it was fixed in HZ 3.5.4 somehow.

Thanks, Denis
